spinc/spinc.go

```diff
@@ -51,155 +46,105 @@ func Run(ctx app.Context) error {
 	}
 
 	// //////////////////////////////////////////////////////////////////////
 	// Request Manager Client
 	// //////////////////////////////////////////////////////////////////////
-	if rmc == nil {
-		var err error
-		rmc, err = makeRMC(&ctx)
-		if err != nil {
-			return err
+	var err error
+	ctx.RMClient, err = makeRMC(ctx)
+	if err != nil {
+		if o.Debug {
+			app.Debug("error making RM client: %s", err)
 		}
-	}
-
-	// //////////////////////////////////////////////////////////////////////
-	// Ping
-	// //////////////////////////////////////////////////////////////////////
-	if o.Ping {
-		if _, err := rmc.RequestList(); err != nil {
-			return fmt.Errorf("Ping failed: %s", err)
+		// All cmds except help and version require an RM client
+		if c.Cmd != "help" && c.Cmd != "version" {
+			return err
 		}
-		fmt.Printf("%s OK\n", o.Addr)
-		return nil
 	}
 
 	// //////////////////////////////////////////////////////////////////////
 	// Commands
 	// //////////////////////////////////////////////////////////////////////
-	cmdFactory := &cmd.DefaultFactory{}
 
-	var err error
-	var run app.Command
-	if ctx.Factories.Command != nil {
-		run, err = ctx.Factories.Command.Make(c.Cmd, ctx)
-		if err != nil {
-			switch err {
-			case cmd.ErrNotExist:
-				if o.Debug {
-					app.Debug("user cmd factory cannot make a %s cmd, trying default factory", c.Cmd)
-				}
-			default:
-				return fmt.Errorf("User command factory error: %s", err)
-			}
-		}
-	}
-	if run == nil {
-		if o.Debug {
-			app.Debug("using default factory to make a %s cmd", c.Cmd)
-		}
-		run, err = cmdFactory.Make(c.Cmd, ctx)
-		if err != nil {
-			switch err {
-			case cmd.ErrNotExist:
-				return fmt.Errorf("Unknown command: %s. Run 'spinc help' to list commands.", c.Cmd)
-			default:
-				return fmt.Errorf("Command factory error: %s", err)
-			}
+	spincCmd, err := ctx.Factories.Command.Make(c.Cmd, ctx)
+	if err != nil {
+		switch err {
+		case cmd.ErrNotExist:
+			return fmt.Errorf("Unknown command: %s. Run 'spinc help' to list commands.", c.Cmd)
+		default:
+			return fmt.Errorf("Command factory error: %s", err)
 		}
 	}
 
-	if err := run.Prepare(); err != nil {
+	// Let command prepare to run. The start command makes heavy use of this.
+	if err := spincCmd.Prepare(); err != nil {
 		if o.Debug {
 			app.Debug("%s Prepare error: %s", c.Cmd, err)
 		}
 		switch err {
-		case config.ErrUnknownRequest:
+		case app.ErrUnknownRequest:
 			reqName := c.Args[0]
 			return fmt.Errorf("Unknown request: %s. Run spinc (no arguments) to list all requests.", reqName)
 		default:
 			return err
 		}
 	}
 
-	if err := run.Run(); err != nil {
-		if o.Debug {
-			app.Debug("%s Run error: %s", c.Cmd, err)
-		}
-		return err
+	err = spincCmd.Run()
+	if o.Debug {
+		app.Debug("%s Run error: %s", c.Cmd, err)
 	}
-
-	return nil
+	return err
 }
 
-func makeRMC(ctx *app.Context) (rm.Client, error) {
+func makeRMC(ctx app.Context) (rm.Client, error) {
 	if ctx.Options.Addr == "" {
```

Comment:
There's a reference to --ping in the error message, and also it should be impossible for Addr to be empty now.

Reply:
There’s a hook that can modify the options, so technically it could unset addr. But it's unclear why it would do that, so I can move the defaults after the hook to ensure it can't shoot itself in the foot.